Translate Chapter 7 to PTBR (#17)
JuliaIntro/JuliaIntroBR.jl: book/src/chap07.asciidoc

[[chap07]]
== Iteração

> Esse capítulo é sobre iteração, que é a habilidade de executar um bloco de argumentos repetidamente. Vimos  um tipo de iteração, usando recursão, em <<recursão>>. Vimos também outro tipo, usando um laço +for+, em <<repetição simples>>. Neste capítulo veremos um outro tipo, usando uma atribuição +while+ (_enquanto_ em Português). Mas primeiro eu tenho que dizer um pouco mais sobre atribuição de variável.

Reviewer: "Neste capítulo veremos um outro tipo, _que usa_ uma _declaração_ +while+ (_enquanto_ em Português)."

(1) A frase que eu sugeri se aproxima mais do jeito que nos expressamos no Brasil.
(2) O termo atribuição em Computação tem um significado específico. Não há atribuição em particular com while. Neste caso, é melhor usar o termo declaração ou comando.

"Neste capítulo veremos um outro tipo, usando uma atribuição +while+ (_enquanto_, em Português)."

Reviewer: Tem erro de digitação na palavra declaração: "Neste capítulo veremos um outro tipo, que usa uma _declração_ +while+ (_enquanto_ em Português)."

> Esse capítulo é sobre iteração, que é a habilidade de executar um bloco de argumentos repetidamente. Vimos  um tipo de iteração, usando recursão, em <<recursão>>. Vimos também outro tipo, usando um laço +for+, em <<repetição simples>>. Neste capítulo veremos um outro tipo, que usa uma declração +while+ (_enquanto_ em Português). Mas primeiro eu tenho que dizer um pouco mais sobre atribuição de variável.

Reviewer: ```suggestion
Esse capítulo é sobre iteração, que é a habilidade de executar um bloco de argumentos repetidamente. Vimos  um tipo de iteração, usando recursão, em <<recursion>>. Vimos também outro tipo, usando um laço +for+, em <<simple_repetition>>. Neste capítulo, veremos um outro tipo, que usa uma declração +while+ (_enquanto_ em Português). Mas primeiro eu tenho que dizer um pouco mais sobre atribuição de variável.
```

> (((iteração)))(((recursão)))(((atribuição for)))

Reviewer: (((_declaração_ for)))

=== Reatribuição

> Você talvez tenha descoberto que é permitido fazer mais de uma atribuição para as mesmas variáveis. Uma nova atribuição faz com que uma variável existente se refira a um novo valor (e pare de se referir ao valor antigo).

Reviewer: ```suggestion
Como você deve ter descoberto, é permitido fazer mais de uma atribuição para as mesmas variáveis. Uma nova atribuição faz com que uma variável existente se refira a um novo valor (e pare de se referir ao valor antigo).
```

> Você talez tenha descoberto, que é permitido fazer mais de uma atribuição para as mesmas variáveis. Uma nova atribuição faz com que uma variável existente se refira a um novo valor (e pare de se referir ao valor antigo).

Reviewer: Você _talvez_ tenha descoberto que é permitido fazer...

(1) Erro de digitação
(2) Tire a vírgula pq a frase está sendo "cortada". A leitura não flui.

[[chap07]]
== Iteração

> Esse capítulo é sobre iteração, que é a habilidade de executar um bloco de argumentos repetidamente. Vimos um tipo de iteração usando recursão em <<recursion>>. Vimos também outro tipo, usando um laço +for+ em <<simple_repetition>>. Neste capítulo, veremos um outro tipo, que usa uma declração +while+ (_enquanto_ em Português). Mas antes temos que falar um pouco mais sobre atribuição de variável.

Reviewer: ```suggestion
Esse capítulo é sobre iteração, que é a habilidade de executar um bloco de argumentos repetidamente. Vimos um tipo de iteração usando recursão em <<recursion>>. Vimos também outro tipo, usando um laço +for+ em <<simple_repetition>>. Neste capítulo, veremos um outro tipo, que usa uma declaração +while+ (_enquanto_ em Português). Mas antes temos que falar um pouco mais sobre atribuição de variável.
```

> (((iteração)))(((recursão)))(((comando for)))

Reviewer: ```suggestion
(((iteração)))(((recursão)))(((declaração for)))
```

7
----

A primeira vez que exibimos +x+, o seu valor é 5; na segunda vez, o seu valor é 7.

[[fig07-1]]
.Diagrama de estado
image::images/fig71.svg[]

> <<fig07-1>> mostra que uma _reatribuição_ se parece com um diagrama de estado

Reviewer: Minha sugestão: "<<fig07-1>> mostra como uma _reatribuição_ funciona por meio de um diagrama de estado."

Reviewer: Falta colocar o ponto final no final da frase.

> <<fig07-1>> mostra como uma _reatribuição_ funciona por meio de um diagrama de estado

Reviewer: ```suggestion
<<fig07-1>> mostra como uma _reatribuição_ funciona por meio de um diagrama de estado.
```

(((reatribuição)))(((diagrama de estado)))

> Neste ponto, quero abordar uma fonte comum de confusão. Como Julia usa o sinal de igual (+=+) para atribuição, é tentador interpretar uma afirmação como +a = b+ como uma proposição matemática de igualdade; isto é, a afirmação de que +a+ e +b+ são iguais. Mas essa interpretação está errada.

Reviewer: Minha sugestão: "Neste ponto, quero abordar um _assunto_ comum de confusão. Como Julia usa o sinal de igual (+=+) para _a_ atribuição..."

Na minha opinião, a frase que eu sugeri se aproxima mais do jeito que nos expressamos no Brasil.

> (((atribuição)))

Reviewer: (((comando de atribuição)))

Reviewer: Me desculpa, aqui eu sugeri algo q já definimos com o Abel. Tem que voltar para "(((atribuição)))"

> (((comando de atribuição)))

Reviewer: ```suggestion
(((atribuição)))
```

> Primeiro, a igualdade é uma relação simétrica e a atribuição não é. Por exemplo, em matemática, se latexmath:[\(a=7\)] então latexmath:[\(7=a\)]. Mas em julia, a atribuição +a=7+ é permitida e +7=a+ não.

Reviewer: "Mas em _Julia_, a atribuição +a=7+ é permitida e +7=a+ não."

> Em matemática, também, uma proposição de igualdade ou é verdadeira ou é falta por todo o tempo. Se latexmath:[\(a=b)\)] agora, então latexmath:[\(a\)] será sempre igual a latexmath:[\(b\)]. Em Julia, uma declração de atribuição pode fazer duas variáveis iguais, mas eles não precisam ficar assim:

Reviewer: _Também na matemática_, uma proposição de igualdade ou é verdadeira ou é _falsa para sempre_. ... Em Julia, um _comando_ de atribuição pode fazer com que duas variáveis sejam iguais, mas eles não precisam ficar assim:

Reviewer: Me desculpa, aqui eu sugeri algo q já definimos com o Abel. Fica assim: "Em Julia, uma _atribuição_ pode fazer com que duas variáveis sejam iguais, mas _elas_ não precisam ficar assim:

> Também na matemática, uma proposição de igualdade ou é verdadeira ou é falsa para sempre. Se latexmath:[\(a=b)\)] agora, então latexmath:[\(a\)] será sempre igual a latexmath:[\(b\)]. Em Julia, um comando de atribuição pode fazer duas variáveis iguais, mas eles não precisam ficar assim:

Reviewer: ```suggestion
Também na matemática, uma proposição de igualdade ou é verdadeira ou é falsa para sempre. Se latexmath:[\(a=b\)] agora, então latexmath:[\(a\)] será sempre igual a latexmath:[\(b\)]. Em Julia, uma atribuição pode fazer duas variáveis iguais, mas elas não precisam ficar assim:
```

7
----

A primeira vez que exibimos +x+, o seu valor é 5; na segunda vez, o seu valor é 7.

[[fig07-1]]
.Diagrama de estado
image::images/fig71.svg[]

<<fig07-1>> mostra como uma _reatribuição_ funciona por meio de um diagrama de estado.
(((reatribuição)))(((diagrama de estado)))

> Neste ponto, quero abordar um assunto comum de confusão. Como o Julia usa o sinal de igual (+=+) para a atribuição, é tentador interpretar uma afirmação como +a = b+ como uma proposição matemática de igualdade; isto é, a afirmação de que +a+ e +b+ são iguais. Mas essa interpretação está errada.

Reviewer 2: ```suggestion
Neste ponto, queremos abordar um assunto comum de confusão. Como o Julia usa o sinal de igual (+=+) para a atribuição, é tentador interpretar uma afirmação como +a = b+ como uma proposição matemática de igualdade; isto é, a afirmação de que +a+ e +b+ são iguais. Mas essa interpretação está errada.
```

(((atribuição)))

Primeiro que, a igualdade é uma relação simétrica e a atribuição não é. Por exemplo, em matemática, se latexmath:[\(a=7\)] então latexmath:[\(7=a\)]. Mas em Julia, a atribuição +a=7+ é permitida e +7=a+ não.

Também na matemática, uma proposição de igualdade ou é verdadeira ou é falsa para sempre. Se latexmath:[\(a=b\)] agora, então latexmath:[\(a\)] será sempre igual a latexmath:[\(b\)]. Em Julia, uma atribuição pode fazer duas variáveis iguais, mas elas não precisam ficar assim:

[source,@julia-repl-test]
----
julia> a = 5
5
julia> b = a    # a e b são iguais
5
julia> a = 3    # a e b não são mais iguais
3
julia> b
5
----

A terceira linha altera o valor de +a+ mas não altera o valor de +b+, portanto eles não são mais iguais.

[WARNING]
====
A reatribuição de variáveis geralmente é útil, mas, você deve usá-la com cuidado. Se os valores das variáveis mudarem com frequência, isso pode dificultar a leitura e a depuração do código.

Não é permitido definir uma função que tenha o mesmo nome de uma variável definida anteriormente.
====

=== Atualizando Variáveis

> Um tipo comum de reatribuição é uma atualização, onde o novo valor da variável depende da antiga.

Reviewer: ```suggestion
Um tipo comum de reatribuição é uma atualização, onde o novo valor da variável depende do antigo.
```

(((atualização)))

[source,@julia-repl-test chap07]
----
julia> x = x + 1
8
----

> Isso significa “pegar o valor atual de +x+, adicionar um, e então atualizar +x+ ao novo valor.”

Reviewer 2: ```suggestion
Isso significa “pegue o valor atual de +x+, adicione um, e então atualize +x+ ao novo valor.”
```

.Diagrama de estado
image::images/fig71.svg[]

<<fig07-1>> mostra como uma _reatribuição_ funciona por meio de um diagrama de estado.
(((reatribuição)))(((diagrama de estado)))

Neste ponto, quero abordar um assunto comum de confusão. Como o Julia usa o sinal de igual (+=+) para a atribuição, é tentador interpretar uma afirmação como +a = b+ como uma proposição matemática de igualdade; isto é, a afirmação de que +a+ e +b+ são iguais. Mas essa interpretação está errada.
(((atribuição)))

> Primeiro que, a igualdade é uma relação simétrica e a atribuição não é. Por exemplo, em matemática, se latexmath:[\(a=7\)] então latexmath:[\(7=a\)]. Mas em Julia, a atribuição +a=7+ é permitida e +7=a+ não.

Reviewer: ```suggestion
Primeiro que a igualdade é uma relação simétrica e a atribuição não é. Por exemplo, em matemática, se latexmath:[\(a=7\)] então latexmath:[\(7=a\)]. Mas em Julia, a atribuição +a=7+ é permitida e +7=a+ não.
```

5
----

A terceira linha altera o valor de +a+ mas não altera o valor de +b+, portanto eles não são mais iguais.

[WARNING]
====

> A reatribuição de variáveis geralmente é útil, mas você deve usá-la com cuidado. Se os valores das variáveis mudarem com frequência, isso poderá dificultar a leitura e o debugging do código.

Reviewer: ```suggestion
A reatribuição das variáveis geralmente é útil, mas você deve usá-la com cuidado. Se os valores das variáveis mudarem com frequência, isso pode dificultar a leitura e a depuração do código.
```

> A reatribuição de variáveis geralmente é útil, mas você deve usá-las com cuidado. Se os valores das variáveis mudarem com frequência, isso poderá dificultar a leitura e a depuração do código.

Reviewer: "... mas você deve _usá-la_ com cuidado. Se os valores das variáveis mudarem com frequência, isso poderá dificultar a leitura e _o debugging_ do código."

> Não é permitido definir uma função que tenha o mesmo nome que uma variável definida anteriormente.

Reviewer: "Não é permitido definir uma função que tenha o mesmo nome _de_ uma variável definida anteriormente."

====

> === Atualizando variáveis

Reviewer: "Atualizando _Variáveis_"

Um tipo comum de reatribuição é uma atualização, onde o novo valor da variável depende da antiga.
(((atualização)))

[source,@julia-repl-test chap07]
----
julia> x = x + 1
8
----

> Isso significa "pegar o valor atual de +x+, adicionar um, e então atualizar +x+ ao novo valor."

Reviewer: Isso significa “pegar o valor atual de +x+, adicionar um, e então atualizar +x+ ao novo valor.”
Mudar as aspas, elas são as aspas curvas.

> Se você tenta atualizar uma variável que não existe, você obterá um erro, porque Julia avalia o lado direito antes de atribuir um valor a ++:

Reviewer: "Se você tenta atualizar uma variável que não existe, _uma mensagem de erro aparecerá_, porque Julia avalia o lado direito antes de atribuir um valor a _+x+_:"

Reviewer: O campo "++" tá vazio. Falta fazer a correção.

> Se você tenta atualizar uma variável que não existe, uma mensagem de erro aparecerá, porque Julia avalia o lado direito antes de atribuir um valor a ++:

Reviewer: ```suggestion
Se você tenta atualizar uma variável que não existe, uma mensagem de erro aparecerá, porque Julia avalia o lado direito antes de atribuir um valor a +x+:
```

> (((UndefVarErro)))(((Erro de Variável Indefinida)))

Reviewer: (((UndefVarError)))

> (((UndefVarError)))(((Erro de Variável Indefinida)))

Reviewer: ```suggestion
(((UndefVarError)))
```

[source,@julia-repl-test]
----
julia> y = y + 1
ERROR: UndefVarError: y not defined
----

> Antes que você possa atualizar uma variável, você tem que inicializar ela, geralmente com uma atribuição simples:

Reviewer: "Antes que você possa atualizar uma variável, você tem que _inicializá-la_, geralmente com uma atribuição simples:"

> Antes que você possa atualizar uma variável, você tem que inicializá-la, geralmente com uma atribuição simples:

Reviewer: ```suggestion
Antes que você possa atualizar uma variável, deve-se inicializá-la geralmente com uma atribuição simples:
```

[WARNING]
====

> A reatribuição de variáveis geralmente é útil, mas, você deve usá-la com cuidado. Se os valores das variáveis mudarem com frequência, isso pode dificultar a leitura e a depuração do código.

Reviewer: ```suggestion
A reatribuição de variáveis geralmente é útil, mas você deve usá-la com cuidado. Se os valores das variáveis mudarem com frequência, isso pode dificultar a leitura e a depuração do código.
```

1
----

> Atualizar uma variável adicionando 1 é chamado de _incremento_; subtraindo 1 é chamado de _decremento_.

Reviewer: Minha sugestão: "Atualizar uma variável adicionando 1 é chamado de _incremento_; _e_ subtraindo 1 é chamado de _decremento_."

Na minha opinião, a alteração dá mais fluidez para o texto.

(((incremento)))(((decremento)))

> === A atribuição +while+

Reviewer: "A _Declaração_ +while+"

> Os computadores são frequentemente usados para automatizar tarefas repetitivas. Repetir tarefas idênticas ou semelhantes sem cometer erros é algo que os computadores fazem bem e as pessoas fazem mal. Em um programa de computador, a repetição também é chamada _iteração_.

Reviewer: "Em um programa de computador, a repetição também é chamada  _de_ _iteração_."

> Os computadores são frequentemente usados para automatizar tarefas repetitivas. Repetir tarefas idênticas ou semelhantes sem cometer erros é algo que os computadores fazem bem e as pessoas fazem mal. Em um programa de computador, a repetição também é chamada de _iteração_.

Reviewer: ```suggestion
Os computadores são frequentemente usados para automatizar tarefas repetitivas. Repetir tarefas idênticas ou semelhantes sem cometer erros é algo que os computadores fazem bem, ao contrário das pessoas. Em um programa de computador, a repetição também é chamada de _iteração_.
```

(((iteração)))

> Já vimos duas funções, +contador_regressivo+ e +imprima_n+, que iteram usando recursão. Como essa iteração é muito comum, Julia fornece recursos para fazer isso fácil. Uma para a atribuição +for+ que vimos em <<Repretição_simples>>. Voltaremos a isso mais tarde.

Reviewer: "Como _a_ iteração é muito comum, Julia _dispõe de_ recursos para _simplificá-la_. Um deles _é_ a _declaração_ +for+ que vimos em <<_Repetição__simples>>. Voltaremos a _este assunto_ mais tarde."

A referência <<simple_repetition>> também foi usado na linha 4. Vc tem que decidir qual é formatação q vc vai usar: <<repetição simples>> ou <<Repetição_simples>>.

Reviewer: "Como a iteração é muito comum, Julia dispõe **_de_** recursos para simplificá-la. Um deles é a declaração +for+ que vimos em <<simple_repetition>>."

> Já vimos duas funções, +contador_regressivo+ e +imprima_n+, que iteram usando recursão. Como a iteração é muito comum, Julia dispõe recursos para simplificá-la. Um deles é a declaração +for+ que vimos em <<simple_repetition>>. Voltaremos a este assunto mais tarde.

Reviewer: ```suggestion
Já vimos duas funções, +contador_regressivo+ e +imprima_n+, que iteram usando recursão. Como a iteração é muito comum, Julia dispõe de recursos para simplificá-la. Um deles é a declaração +for+ que vimos em <<simple_repetition>>. Voltaremos a este assunto mais tarde.
```

> (((recursão)))(((atribuição for)))

Reviewer: (((declaração for)))

Já dá um replace e substitui "atribuição for" por "declaração for".

> A outra é para a _atribuição_ +while+. Aqui está a versão do +contagem_regressiva+ que utiliza a atribuição +while+:

Reviewer: "O outro recurso é a _declaração_ +while+. Aqui está a versão do +contagem_regressiva+ que utiliza a _declaração_ +while+:"

Já dá um replace e substitui "atribuição while" por "declaração while".

> A outra é para a _declaração_ +while+. Aqui está a versão do +contagem_regressiva+ que utiliza a declaração +while+:

Reviewer: ```suggestion
O outro é a _declaração_ +while+. Aqui está a versão do +contagem_regressiva+ que utiliza a declaração +while+:
```

> (((while)))((("keyword", "while", see="while")))(((atribuição while)))((("atribuição", "while", see="atribuição while")))

Reviewer: (((while)))(((_"palavra-chave_", "while", see="while")))(((_declaração while_)))((("declaração", "while", see="_declaração while_")))

Reviewer: Ainda tem correções nessa linha 104.

> (((while)))((("keyword", "while", see="while")))(((declaração while)))((("atribuição", "while", see="declaração while")))

Reviewer: ```suggestion
(((while)))((("palavra-chave", "while", see="while")))(((declaração while)))((("declaração", "while", see="declaração while")))
```

[source,@julia-setup]
----
function contagem_regressiva(n)
    while n > 0
        print(n, " ")
        n = n - 1
    end
    println("Vai!")
end
----

> Você pode ler quase toda a atribuição +while+ como se estivesse em inglês (ou _Português_). Isso significa que, "Enquanto +n+ for maior que 0, exiba o valor de +n+ e decrementa +n+. Quando atingir 0, exiba a palavra Vai!"

Reviewer: "Isso significa que "Enquanto +n+ for maior que 0, exiba o valor de +n+ e _diminua_ +n+. Quando _chegar a_ 0, _imprima_ a palavra Vai!"

Reviewer: Ainda tem que fazer alteração no termo "atribuição while".

"Você pode ler quase toda a **_declaração_** +while+ como se estivesse em inglês (ou _Português_). Isso significa que "Enquanto +n+ for maior que 0, exiba o valor de +n+ e diminua +n+. Quando chegar a 0, imprima a palavra Vai!"

> Você pode ler quase toda a atribuição +while+ como se estivesse em inglês (ou _Português_). Isso significa que "Enquanto +n+ for maior que 0, exiba o valor de +n+ e diminua +n+. Quando chegar a 0, imprima a palavra Vai!"

Reviewer: ```suggestion
Você pode ler quase toda a declaração +while+ como se estivesse em inglês. Isso significa que "Enquanto +n+ for maior que 0, exiba o valor de +n+ e depois diminua +n+. Quando chegar a 0, imprima a palavra Vai!"
```

> Você pode ler quase toda a declaração +while+ como se estivesse em inglês. Isso significa que “Enquanto +n+ for maior que 0, exiba o valor de +n+ e depois diminua +n+. Quando chegar a 0, imprima a palavra Decolar!”

Reviewer 2: ```suggestion
Você pode ler a declaração +while+ quase como se estivesse em Português, traduzindo _while_ por _enquanto_. Ela significa “Enquanto +n+ for maior que 0, exiba o valor de +n+ e depois diminua +n+. Quando chegar a 0, imprima a palavra Decolar!”
```

> (((contagem regressiva)))

Reviewer: Eu acho q esta referência é para a função e neste caso, ficaria (((contagem_regressiva))).

> Formalmente, aqui está o fluxo de execução de uma atribuição +while+:

Reviewer: ```suggestion
Formalmente, aqui está o fluxo de execução de uma declaração +while+:
```

(((fluxo de execução)))

. Determinar se a condição é verdadeira ou falsa.

> . Se for falsa, saia do comando _while_ e continue a execução para a próxima instrução.

Reviewer: ```suggestion
. Se for falsa, saia do comando _while_ e continue a execução para o próximo comando.
```

> . Se for falsa, saia da atribuição _while_ e continue a execução para o próximo comando.

Reviewer: ```suggestion
. Se for falsa, saia da declaração _while_ e continue a execução para o próximo comando.
```

> . Se a condição for verdadeira, execute os comandos e então volte para o passo 1.

Reviewer: ```suggestion
. Se a condição for verdadeira, execute os comandos do corpo e então volte para o passo 1.
```

Esse tipo de fluxo é chamado de laço porque o terceiro passo retorna ao topo.
(((laço)))

> O corpo do laço deve mudar o valor de uma ou mais variáveis até que a condição eventualmente se torna falsa e o laço é finalizado. Caso contrário, o laço irá se repetir para sempre, que é um _laço infinito_. Uma fonte inesgotável de diversão para os cientistas da computação são as instruções nos frascos de shampoo, “ensaboar, enxaguar, repetir”, que é um laço infinito.

Reviewer: ```suggestion
O corpo do laço deve mudar o valor de uma ou mais variáveis até que eventualmente a condição se torna falsa e o laço é finalizado. Caso contrário, o laço irá se repetir para sempre, que é um _laço infinito_. Uma fonte inesgotável de diversão para os cientistas da computação são as instruções nos frascos de shampoo, “ensaboar, enxaguar, repetir”, que é um laço infinito.
```

> O corpo do laço pode mudar o valor de uma ou mais variáveis até que então a condição eventualmente torna falsa e o laço é finalizado. Caso contrário, o laço irá se repetir pra sempre, o que é um _laço infinito_. Uma fonte inesgotável de diversão para os cientistas da computação é a observação de que as instruções nos frascos de shampoo, “ensaboar, enxaguar, repetir”, são um laço infinito.

Reviewer: "O corpo do laço _deve_ mudar o valor de uma ou mais variáveis até que a condição eventualmente se torna falsa e o laço é finalizado. Caso contrário, o laço irá se repetir _para_ sempre, que é um _laço infinito_. Uma fonte inesgotável de diversão para os cientistas da computação _são_ as instruções nos frascos de shampoo, “ensaboar, enxaguar, repetir”, que é um laço infinito."

> O corpo do laço deve mudar o valor de uma ou mais variáveis até que eventualmente a condição se torna falsa e o laço é finalizado. Caso contrário, o laço irá se repetir para sempre, ou seja, um _laço infinito_. Uma fonte inesgotável de diversão para os cientistas da computação são as instruções nos frascos de shampoo, “ensaboar, enxaguar, repetir”, pois é um laço infinito.

Reviewer 2: ```suggestion
O corpo do laço deve mudar o valor de uma ou mais variáveis até que eventualmente a condição se torna falsa e o laço é finalizado. Caso contrário, o laço irá se repetir para sempre, ou seja, um _laço infinito_. Uma fonte inesgotável de diversão para os cientistas da computação são as instruções nos frascos de shampoo, “ensaboe, enxague, repita”, pois é um laço infinito.
```

(((laço infinito)))

> No caso da +contagem_regressiva+, podemos provar que o laço é finalizado: se +n+ é zero ou negativo, o laço nunca termina. Caso contrário, +n+ vai diminuindo através do laço, e então eventualmente chegará a 0.

Reviewer: ```suggestion
No caso da +contagem_regressiva+, podemos provar que o laço é finalizado: se +n+ é zero ou negativo, o laço nunca termina. Caso contrário, +n+ vai diminuindo durante o laço, e então eventualmente chegará a 0.
```

> No caso da +contagem_regressiva+, podemos provar que o laço é finalizado: se +n+ é zero ou negativo, o laço nunca termina. Caso contrário, +n+ vai atingindo valores menores através do laço, então eventualmente chegará a 0.

Reviewer: "Caso contrário, +n+ vai _diminuindo_ através do laço, _e_ então eventualmente chegará a 0."

> Para alguns laços, não é fácil provar. Por exemplo:

Reviewer: "Para alguns laços, não é _tão_ fácil de provar."

(((seq)))((("função", "definido pelo programador", "seq", see="seq")))

[source,@julia-setup]
----
function seq(n)
    while n != 1
        println(n)
        if n % 2 == 0        # n é par
            n = n / 2
        else                 # n é ímpar
            n = n*3 + 1
        end
    end
end
----

> A condição para esse laço é +n != 1+, então esse laço irá continuar até que +n+ atinja 1, o que faz a condição ser falsa.

Reviewer: ```suggestion
A condição para esse laço é +n != 1+, então esse laço irá continuar até que +n+ seja 1, o que faz a condição ser falsa.
```

> A cada turno do laço, o programa tem como saída o valor +n+ que é verificado se é par ou ímpar. Se for par, +n+ é dividido por 2. Se é ímpar, o valor de +n+ é substituído por +pass:[n*3 + 1]+. Por exemplo, se o argumento da sequência é 3, os valores de +n+ recebidos são 3, 10, 5, 16, 8, 4, 2, 1.

Reviewer: ```suggestion
A cada turno do laço, o programa tem como saída o valor +n+ que é verificado se é par ou ímpar. Se é par, +n+ é dividido por 2. Se é ímpar, o valor de +n+ é substituído por +pass:[n*3 + 1]+. Por exemplo, se o argumento da sequência é 3, os valores de +n+ recebidos são 3, 10, 5, 16, 8, 4, 2, 1.
```

> A cada turno do laço, o programa tem como saída o valor +n+ que é verificado se é par ou ímpar. Se for par, +n+ é divisível por 2. Se é ímpar, o valor de +n+ é substituído por +pass:[n*3 + 1]+. Por exemplo, se o argumento da sequência é 3, os valores de +n+ recebidos são 3, 10, 5, 16, 8, 4, 2, 1.

Reviewer: ```suggestion
A cada turno do laço, o programa tem como saída o valor +n+ que é verificado se é par ou ímpar. Se for par, +n+ é dividido por 2. Se é ímpar, o valor de +n+ é substituído por +pass:[n*3 + 1]+. Por exemplo, se o argumento da sequência é 3, os valores de +n+ recebidos são 3, 10, 5, 16, 8, 4, 2, 1.
```

> Cada vez através do laço, o programa tem como saída o valor +n+ e verifica se é par ou ímpar. Se for par, +n+ é divisível por 2. Se é ímpar, o valor de +n+ é substituído por +pass:[n*3 + 1]+. Por exemplo, se o argumento da sequência é 3, o valores do resultado de +n+ são 3, 10, 5, 16, 8, 4, 2, 1.

Reviewer: "_A cada turno_ do laço, o programa tem como saída o valor +n+ _que é verificado_ se é par ou ímpar. Se for par, +n+ é _dividido_ por 2. Se é ímpar, o valor de +n+ é substituído por +pass:[n*3 + 1]+. Por exemplo, se o argumento _passado para a_ sequência é 3, _os valores de +n+ recebidos_ são 3, 10, 5, 16, 8, 4, 2, 1."

> A cada turno do laço, o programa tem como saída o valor +n+ que é verificado se é par ou ímpar. Se é par, +n+ é dividido por 2. Se é ímpar, o valor de +n+ é substituído por +pass:[n*3 + 1]+. Por exemplo, se o argumento da sequência é 3, os valores de +n+ recebidos são 3, 10, 5, 16, 8, 4, 2, 1.

Reviewer 2: ```suggestion
A cada passada do laço, o programa tem como saída o valor +n+ que é verificado se é par ou ímpar. Se é par, +n+ é dividido por 2. Se é ímpar, o valor de +n+ é substituído por +pass:[n*3 + 1]+. Por exemplo, se o argumento da sequência é 3, os valores de +n+ recebidos são 3, 10, 5, 16, 8, 4, 2, 1.
```

> Já que +n+ as vezes cresça e as vezes decresça, não existe uma demonstração óbvia de que +n+ alcançará 1, ou que o programa termine. Para alguns valores particulares de +n+, podemos demonstrar que termina. Por exemplo, se o valor inicial é uma potência de dois, +n+ será sempre par através do laço até que atinja 1. O exemplo anterior finaliza essa sequência, a partir de 16.

Reviewer: ```suggestion
Já que +n+ às vezes cresce e às vezes decresce, não existe uma demonstração óbvia de que +n+ resultará em 1, ou que o programa termine. Para alguns valores particulares de +n+, podemos demonstrar que termina. Por exemplo, se o valor inicial é uma potência de dois, +n+ será sempre par durante o laço até que resulta em 1. O exemplo anterior finaliza essa sequência, a partir de 16.
```

> Já que +n+ às vezes cresce e às vezes decresce, não existe uma demonstração óbvia de que +n+ resultará em 1 ou que o programa termine. Para alguns valores particulares de +n+, podemos demonstrar que termina. Por exemplo, se o valor inicial é uma potência de dois, +n+ será sempre par durante o laço até que resulta em 1. O exemplo anterior finaliza essa sequência, a partir de 16.

Reviewer: ```suggestion
Já que +n+ às vezes cresce e às vezes decresce, não existe uma demonstração óbvia de que +n+ resultará em 1 ou que o programa termine. Para alguns valores particulares de +n+, podemos demonstrar que termina. Por exemplo, se o valor inicial é uma potência de dois, +n+ será sempre par durante o laço até que chega em 1. O exemplo anterior finaliza essa sequência, a partir de 16.
```

> Desde que +n+ as vezes cresça e as vezes decresça, não existe uma demonstração óbvia de que +n+ alcançará 1, o que o programa termine. Para alguns valores particulares de +n+, podemos demonstrar que termina. Por exemplo, se o valor inicial é uma potência de dois, +n+ será sempre par através do laço até que atinja 1. O exemplo anterior finaliza com essa sequência, começando com 16.

Reviewer: "_Já que +n+ às vezes cresce e às vezes decresce_, não existe uma demonstração óbvia de que +n+ alcançará 1, _ou que o programa termine_. Para alguns valores específicos de +n+, podemos demonstrar que termina. Por exemplo, se o valor inicial é uma potência de dois, +n+ será sempre par _durante o_ laço até que atinja 1. O exemplo anterior finaliza _essa sequência_, a partir de 16."

Reviewer: Ainda tem correções nessa linha:
"Já que +n+ **_às vezes cresce e às vezes decresce_**, não existe uma demonstração óbvia de que +n+ alcançará 1, ou que o programa termine. Para alguns valores **_específicos_** de +n+, podemos demonstrar que termina. Por exemplo, se o valor inicial é uma potência de dois, +n+ será sempre par durante o laço até que **_chega a_** 1. O exemplo anterior finaliza essa sequência, a partir de 16."

> A parte difícil é provar que esse programa finaliza para todos os valores positivos de +n+. Até agora ninguém foi capaz de provar ou desprovar isso! (consulte https://pt.wikipedia.org/wiki/Conjectura_de_Collatz.)

Reviewer: ```suggestion
A parte difícil é provar que esse programa finaliza para todos os valores positivos de +n+. Até agora ninguém foi capaz de provar ou desprovar isso! (Consulte https://pt.wikipedia.org/wiki/Conjectura_de_Collatz.)
```

> A parte difícil é se podemos provar que esse programa finaliza para todos os valores positivos de +n+. Portanto, até agora ninguém foi capaz de provar ou desprovar isso! (consulte https://en.wikipedia.org/wiki/Collatz_conjecture.)

Reviewer: ```suggestion
A parte difícil é se podemos provar que esse programa finaliza para todos os valores positivos de +n+. Até agora ninguém foi capaz de provar ou desprovar isso! (consulte https://pt.wikipedia.org/wiki/Conjectura_de_Collatz.)
```

> A parte difícil é se podemos provar que esse programa finaliza para todos os valores positivos de +n+. Portanto, ninguém foi capaz de provar ou desprovar isso! (see https://en.wikipedia.org/wiki/Collatz_conjecture.)

Reviewer: "_Até agora_ ninguém foi capaz de provar ou desprovar isso! (_consulte_ https://pt.wikipedia.org/wiki/Conjectura_de_Collatz.)"

Reviewer: Tb tem q alterar o link do wikipedia. O link q eu coloquei no comentário é em português.

(((Conjectura de Collatz)))

===== Exercício 7-1

> Reescreva a função +imprima_n+ de <<Recursão>> usando iteração ao invés de recursão.

Reviewer: ```suggestion
Reescreva a função +imprima_n+ de <<recursion>> usando iteração ao invés de recursão.
```

=== +break+

> Às vezes, você não sabe que é hora de terminar um laço até chegar na metade do corpo. Neste caso você pode utilizar o _comando break_ para sair do laço.

Reviewer: ```suggestion
Às vezes, você não sabe que é hora de terminar um laço até chegar na metade do corpo. Neste caso você pode utilizar a _declaração break_ para sair do laço.
```

> Às vezes, você não sabe que é hora de terminar um laço até chegar na metade do corpo. Neste caso você pode utilizar a _atribuição de parada_ para sair do laço.

Reviewer: "Neste caso você pode utilizar a _declaração break_ para sair do laço."

> (((break)))((("palavra-chave", "break", see="break")))(((comando break)))((("comando", "break", see="comando break")))

Reviewer: ```suggestion
(((break)))((("palavra-chave", "break", see="break")))(((declaração break)))((("declaração", "break", see="declaração break")))
```

> (((break)))((("palavra-chave", "parada", see="parada")))(((declaração break)))((("atribuição", "parada", see="declaração break")))

Reviewer: ```suggestion
(((break)))((("palavra-chave", "break", see="break")))(((declaração break)))((("declaração", "break", see="declaração break")))
```

> (((break)))((("palavra chave", "parada", see="parada")))(((atribuição de parada)))((("atribuição", "parada", see="atribuição de parada")))

Reviewer: ((("palavra-chave", "break", see="break")))(((declaração break)))((("declaração", "break", see="declaração break")))

Estamos fazendo referência ao comando break, que é uma palavra-chave do Julia, não sobre o assunto de parar um laço de repetição.

Reviewer: Ainda tem correção na linha 171.

((("palavra-chave", "break", see="break"))) precisa ficar assim porque break é uma palavra-chave do Julia e não pode ser usado como nome de variável.

O último conjunto de triplas aspas "((("declaração", "break", see="declaração break")))" também tem que ficar assim porque está falando da declaração break, que não é uma atribuição e reforço o que eu disse no comentário anteior, estamos falando da declaração break no Julia e não necessariamente do que o break faz em um laço. Na tradução em espanhol, ficou assim:
"(((break)))((("palabra reservada", "break", see="break")))(((sentencia break)))((("sentencia", "break", see="sentencia break")))"

> Por exemplo, suponha que você quer receber entradas do usuário até que ele digite concluído. Poderia-se escrever:

Reviewer: ```suggestion
Por exemplo, suponha que você queira receber entradas do usuário até que ele digite concluído. Poderia-se escrever:
```

> Por exemplo, suponha que você quer receber entradas do usuário até que ele digite concluído. Você poderia escrever:

Reviewer: ```suggestion
Por exemplo, suponha que você quer receber entradas do usuário até que ele digite concluído. Poderia-se escrever:
```

> Por exemplo, suponha que você quer coletar entradas do usuário até que ele digite concluído. Você poderia escrever:

Reviewer: "Por exemplo, suponha que você quer _receber as_ entradas do usuário até que ele digite concluído."

> ((leitura de linha))

Reviewer: (((readline)))

Estamos fazendo referência ao comando readline usada na função mostrada no exemplo.

[source,julia]
----
while true
    print("> ")

>     line = readline()

Reviewer: linha = readline()

Também estou traduzindo, qdo possível, partes do código.

>     if line == "concluído"

Reviewer: if linha == "concluído"

        break
    end

>     println(line)

Reviewer: println(linha)

end
println("Concluído!")
----

> A condição deste laço é +true+, que é sempre verdade, então o laço será executado até chegar no comando break.

Reviewer: ```suggestion
A condição deste laço é +true+, que é sempre verdade, então o laço será executado até chegar na declaração break.
```

> A condição deste laço é +true+, o que ocorre sempre, então o laço irá ser executado até chegar na declaração break.

Reviewer: ```suggestion
A condição deste laço é +true+, que é sempre verdade, então o laço será executado até chegar na declaração break.
```

> A condição de laço é +true+, o que ocorre sempre, então o laço irá ser executado até que alcance a atribuição de pausa.

Reviewer: "A condição _deste_ laço é +true+, que é sempre verdadeiro, então o laço será interrompido somente quando chegar na declaração break."

> A cada iteração, a solicitação ao usuário ocorre por meio de um sinal de maior (">"). Se o usuário digitar +concluído+, então o comando break finaliza o laço. Caso contrário, o programa mostrará o que o usuário digitar e voltará ao topo do laço. Aqui está um exemplo de execução:

Reviewer: ```suggestion
A cada iteração, a solicitação ao usuário ocorre por meio de um sinal de maior (">"). Se o usuário digitar +concluído+, então a declaração break finaliza o laço. Caso contrário, o programa mostrará o que o usuário digitar e voltará ao topo do laço. Aqui está um exemplo de execução:
```

> A cada vez, ele solicita ao usuário um colchete angular. Se o usuário digitar +concluído+, a instrução break sai do laço. Caso contrário, o programa ecoará o que o usuário digitar e voltará ao topo do laço. Aqui está uma amostra de execução:

Reviewer: "A cada iteração, a solicitação ao usuário ocorre por meio de um sinal de maior (">") . Se o usuário digitar +concluído+, então o comando break finaliza o laço. Caso contrário, o programa mostrará o que o usuário digitar e voltará ao topo do laço. Aqui está uma mostra da execução:"

> A cada iteração, a solicitação ao usuário ocorre por meio de um sinal de maior (">"). Se o usuário digitar +concluído+, então o comando break finaliza o laço. Caso contrário, o programa mostrará o que o usuário digitar e voltará ao topo do laço. Aqui está uma mostra de execução:

Reviewer: ```suggestion
A cada iteração, a solicitação ao usuário ocorre por meio de um sinal de maior (">"). Se o usuário digitar +concluído+, então a declaração break finaliza o laço. Caso contrário, o programa mostrará o que o usuário digitar e voltará ao topo do laço. Aqui está uma mostra de execução:
```

[source]
----
> não está concluído
não está concluído
> concluído
Concluído!
----

> Essa maneira de escrever laços é comum porque você pode verificar a condição em qualquer lugar do laço (não apenas no topo) e você pode expressar a condição de parada afirmativamente ("pare quando isso acontecer") tanto quanto negativamente ("continue enquanto isso acontece").

Reviewer: ```suggestion
Essa maneira de escrever laços é comum porque você pode verificar a condição em qualquer lugar do laço (não apenas no topo) e você pode expressar a condição de parada afirmativamente ("pare quando isso acontecer") ao invés de negativamente ("continue enquanto isso acontece").
```

=== +continue+

> O comando break sai do laço. Quando um _comando continue_ é encontrado dentro de um laço, salta-se para o início do laço da próxima iteração, pulando a execução de comandos dentro do corpo do laço da iteração atual. Por exemplo:

Reviewer: ```suggestion
A declaração break sai do laço. Quando uma _declaração continue_ é encontrado dentro de um laço, salta-se para o início do laço da próxima iteração, pulando a execução de comandos dentro do corpo do laço da iteração atual. Por exemplo:
```

> A atribuição de parada sai do laço. Quando uma _atribuição de continuidade_ é encontrado dentro de um laço, o controle salta para o início do laço para a próxima iteração, pulando a execução de instruções dentro do corpo do laço para a iteração atual. Por exemplo:

Reviewer: "A declaração break sai do laço. Quando uma _declaração continue_ é encontrada dentro de um laço, salta-se para o início do laço da próxima iteração, pulando a execução de comandos dentro do corpo do laço da iteração atual. Por exemplo:"

Reviewer: Ainda tem correções na linha 206:

"A declaração break sai do laço. Quando uma _declaração continue_ é **_encontrada_** dentro de um laço, salta-se para o início do laço _**da**_ próxima iteração, pulando a execução de comandos dentro do corpo do laço **_da_** iteração atual. Por exemplo:"

> A declaração break sai do laço. Quando uma _declaração continue_ é encontrado dentro de um laço, salta-se para o início do laço para a próxima iteração, pulando a execução de comandos dentro do corpo do laço para a iteração atual. Por exemplo:

Reviewer: ```suggestion
A declaração break sai do laço. Quando uma _declaração continue_ é encontrada dentro de um laço, salta-se para o início do laço da próxima iteração, pulando a execução de comandos dentro do corpo do laço da iteração atual. Por exemplo:
```

> (((continue)))((("palavra-chave", "continue", see="continue")))(((comando continue)))((("comando", "continue", see="continue", see="comando continue")))

Reviewer: ```suggestion
(((continue)))((("palavra-chave", "continue", see="continue")))(((declaração continue)))((("declaração", "continue", see="continue", see="declaração continue")))
```

> (((continue)))((("palavra-chave", "continue", see="continue")))(((declaração continue)))((("atribuição", "continuar", see="continuar", see="declaração continue")))

Reviewer: ```suggestion
(((continue)))((("palavra-chave", "continue", see="continue")))(((declaração continue)))((("declaração", "continue", see="continue", see="declaração continue")))
```

> (((continue)))((("palavra-chave", "continuar", see="continuar")))(((atribuição de continuidade)))((("atribuição", "continuar", see="continuar", see="atribuição de continuidade")))

Reviewer: ((("palavra-chave", "continue", see="continue")))(((declaração continue)))((("declaração", "continue", see="continue", see="declaração continue")))

(1) na linha 171, vc escreve "palavra chave", sem hífen. Vc precisa padronizar a escrita de _keyword_.
(2) esta referência é para a keyword continue, não sobre a definição de um comando sobre continuar.

Reviewer: Falta essa correção: "((("declaração", "continue", see="continue", see="declaração continue")))"

(((incremento)))(((decremento)))

> === O Comando +while+

Reviewer: ```suggestion
=== A Declaração +while+
```

Os computadores são frequentemente usados para automatizar tarefas repetitivas. Repetir tarefas idênticas ou semelhantes sem cometer erros é algo que os computadores fazem bem, ao contrário das pessoas. Em um programa de computador, a repetição também é chamada de _iteração_.
(((iteração)))

> Já vimos duas funções, +contador_regressivo+ e +imprima_n+, que iteram usando recursão. Como a iteração é muito comum, Julia dispõe de recursos para simplificá-la. Um deles é o comando +for+ que vimos em <<simple_repetition>>. Voltaremos a este assunto mais tarde.

Reviewer: ```suggestion
Já vimos duas funções, +contador_regressivo+ e +imprima_n+, que iteram usando recursão. Como a iteração é muito comum, Julia dispõe de recursos para simplificá-la. Um deles é a declaração +for+ que vimos em <<simple_repetition>>. Voltaremos a este assunto mais tarde.
```

> (((recursão)))(((comando for)))

Reviewer: ```suggestion
(((recursão)))(((declaração for)))
```

> O outro é a _comando_ +while+. Aqui está a versão do +contagem_regressiva+ que utiliza o comando +while+:

Reviewer: ```suggestion
A outra é a _declaração_ +while+. Aqui está a versão de +contagem_regressiva+ que utiliza a declaração +while+:
```

> (((while)))((("palavra-chave", "while", see="while")))(((comando while)))((("comando", "while", see="comando while")))

Reviewer: ```suggestion
(((while)))((("palavra-chave", "while", see="while")))(((declaração while)))((("declaração", "while", see="declaraçao while")))
```

        print(n, " ")
        n = n - 1
    end

>     println("Feliz Ano Novo!")

Reviewer: ```suggestion
    println("Decolar!")
```

end
----

> Você pode ler quase toda o comando +while+ como se estivesse em inglês. Isso significa que "Enquanto +n+ for maior que 0, exiba o valor de +n+ e depois diminua +n+. Quando chegar a 0, imprima a a frase Feliz Ano Novo!"

Reviewer: ```suggestion
Você pode ler quase toda a declaração +while+ como se estivesse em inglês. Isso significa que “Enquanto +n+ for maior que 0, exiba o valor de +n+ e depois diminua +n+. Quando chegar a 0, imprima a palavra Decolar!”
```

(((contagem_regressiva)))

> Formalmente, segue abaixo o fluxo de execução de um comando +while+:

Reviewer: ```suggestion
Formalmente, segue abaixo o fluxo de execução de uma declaração +while+:
```

end
----

> Se +i+ é divisível por 3, o comando continue para na iteração atual e a próxima iteração é iniciada. Apenas os números no intervalo entre 1 a 10 não divisíveis por 3 são exibidos.

Reviewer: ```suggestion
Se +i+ é divisível por 3, a declaração continue para na iteração atual e a próxima iteração é iniciada. Apenas os números no intervalo entre 1 a 10 não divisíveis por 3 são exibidos.
```

> Se +i+ é divisível por 3, a atribuição de continuidade para na iteração atual e a próxima iteração é iniciada. Apenas os números no intervalo entre 1 a 10 não divisíveis por 3 são exibidos.

Reviewer: "Se +i+ é divisível por 3, a _declaração continue_ para na iteração atual e a próxima iteração é iniciada. Apenas os números no intervalo entre 1 a 10 não divisíveis por 3 são exibidos."

> [[raizes_quadradas]]

Reviewer: ```suggestion
[[square_roots]]
```

=== Raízes Quadradas

> Laços são geralmente usados em programas que computam resultados numéricos começando com um valor aproximado e é refinado iterativamente.

Reviewer: ```suggestion
Laços são frequentemente usados em programas que calculam resultados numéricos começando com um valor aproximado e aprimorando-o iterativamente.
```

> Laços são geralmente usados em programas que computam resultados numéricos começando com uma resposta aproximada e iterativamente melhorando.

Reviewer: Minha sugestão: "... que computam resultados numéricos começando com _um valor aproximado_ e _é refinado iterativamente_."

Na minha opinião, a frase q eu sugeri se aproxima mais do jeito que nos expressamos no Brasil.

> Por exemplo, uma maneira de computar raízes quadradas é através do Método de Newton. Suponha que você queira saber a raíz quadrada de latexmath:[\(a\)]. Se você começar com uma estimativa qualquer, latexmath:[\(x\)], pode-se calcular uma estimativa melhor com a seguinte fórmula:

Reviewer: ```suggestion
Por exemplo, uma maneira de computar raízes quadradas é através do método de Newton. Suponha que você queira saber a raíz quadrada de latexmath:[\(a\)]. Se você começar com uma estimativa qualquer, latexmath:[\(x\)], pode-se calcular uma estimativa melhor com a seguinte fórmula:
```

> Por exemplo, uma maneira de computar raízes quadradas é através do método de Newton. Suponha que você queira saber a raíz quadrada de latexmath:[\(a\)]. Se você começar com uma estimativa qualquer, latexmath:[\(x\)], pode-se calcular uma estimativa melhor com a seguinte fórmula:

Reviewer: ```suggestion
Por exemplo, uma maneira de computar raízes quadradas é através do método de Newton. Suponha que você queira saber a raiz quadrada de latexmath:[\(a\)]. Se você começar com uma estimativa qualquer, latexmath:[\(x\)], pode-se calcular uma estimativa melhor com a seguinte fórmula:
```

> Por exemplo, uma maneira de computar raízes quadradas é através do método de Newton. Suponha que você queira saber a raiz quadrada de latexmath:[\(a\)]. Se você começar com uma estimativa qualquer, latexmath:[\(x\)], pode-se calcular uma estimativa melhor com a seguinte fórmula:

Reviewer 2: ```suggestion
Por exemplo, uma maneira de calcular raízes quadradas é através do método de Newton. Suponha que você queira saber a raiz quadrada de latexmath:[\(a\)]. Se você começar com uma estimativa qualquer, latexmath:[\(x\)], pode-se calcular uma estimativa melhor com a seguinte fórmula:
```

> Por exemplo, um jeito de computar raízes quadradas é através do Método de Newton. Suponha que você quer saber a raíz quadrada de latexmath:[\(a\)]. Se você começar com quase qualquer uma estimativa, latexmath:[\(x\)], pode-se computar uma estimativa melhor com a seguinte fórmula:

Reviewer: ```suggestion
Por exemplo, um jeito de computar raízes quadradas é através do Método de Newton. Suponha que você quer saber a raiz quadrada de latexmath:[\(a\)]. Se você começar com uma estimativa qualquer, latexmath:[\(x\)], pode-se calcular uma estimativa melhor com a seguinte fórmula:
```

> (((Método de Newton)))

Reviewer: ```suggestion
(((método de Newton)))
```

2.1666666666666665
----

> O resultado é próximo da resposta correta (latexmath:[\(\sqrt 4 = 2\)]). Se repetirmos o processo com a nova estimativa, irá ficar mais próximo ainda:

Reviewer: "O resultado _está mais_ próximo da resposta correta (latexmath:[\(\sqrt 4 = 2\)]). Se repetirmos o processo com a nova estimativa, irá ficar mais próximo ainda:"

> O resultado está mais próximo da resposta correta (latexmath:[\(\sqrt 4 = 2\)]). Se repetirmos o processo com a nova estimativa, irá ficar mais próximo ainda:

Reviewer: ```suggestion
O resultado está mais próximo da resposta correta (latexmath:[\(\sqrt 4 = 2\)]). Se repetirmos o processo com a nova estimativa, ficará mais próximo ainda:
```

2.0064102564102564
----

> Depois de algumas atualizações, a estimativa é quase exata:

Reviewer: Depois de _mais_ algumas atualizações, a estimativa é quase exata:

> Depois de mais algumas atualizações a estimativa é quase exata:

Reviewer: ```suggestion
Depois de mais algumas atualizações, a estimativa é quase exata:
```

2.0000000000262146
----

> Em geral, não sabemos antecipadamente quantas etapas são necessárias para obter a resposta certa, mas sabemos quando chegamos lá porque a estimativa para de mudar:

Reviewer: ```suggestion
Em geral, não sabemos antecipadamente quantos passos são necessários para obter a resposta certa, mas sabemos quando chegamos lá porque a estimativa para de mudar:
```

2.0
----

> 

Reviewer: A linha 271 não foi traduzida.

> Quando +y == x+, podemos parar. Aqui está um laço que começa com uma estimativa inicial, +x+, e melhora até parar de mudar

Reviewer: ```suggestion
Quando +y == x+, podemos parar. Aqui está um laço que começa com uma estimativa inicial +x+, e melhora até parar de mudar:
```

> Quando +y == x+, podemos parar. Aqui está um laço que começa com uma estimativa inicial, +x+, e melhora até que parar de mudar:

Reviewer: "Aqui está um laço que começa com uma estimativa inicial, +x+, e melhora até parar de mudar:"

end
----

> Para a maior parte dos valores de a funciona bem, mas em geral é perigo testar igualdade de float. Pontos Flutuantes são apenas aproximadamente corretos: números mais racionais, como latexmath:[\(\frac{1}{3}\)], e números irracionais como latexmath:[\(\sqrt 2\)], não podem ser representados exatamente com um +Float64+.

Reviewer: "Esta função funciona bem para a maior parte dos valores de a, mas em geral é perigoso testar igualdade com pontos flutuantes. Pontos flutuantes não são totalmente corretos: a maioria dos números racionais como latexmath:[\(\frac{1}{3}\)], e números irracionais como latexmath:[\(\sqrt 2\)], não podem ser representados exatamente com um tipo +Float64+."

> Esta função funciona bem para a maior parte dos valores de a, mas em geral é perigoso testar igualdade com pontos flutuantes. Pontos flutuantes não são totalmente corretos: a maioria dos números racionais como latexmath:[(\frac{1}{3})], e números irracionais como latexmath:[(\sqrt 2)], não podem ser representados exatamente com um tipo +Float64+.

Reviewer: ```suggestion
Esta função funciona bem para a maior parte dos valores de a, mas em geral é perigoso testar igualdade com pontos flutuantes. Pontos flutuantes não são totalmente corretos: a maioria dos números racionais, como latexmath:[(\frac{1}{3})], e números irracionais, como latexmath:[(\sqrt 2)], não podem ser representados exatamente com um tipo +Float64+.
```

Em vez de verificar se +x+ e +y+ são exatamente iguais, é mais seguro usar a função interna +abs+ para calcular o valor absoluto, ou magnitude, da diferença entre eles:

> (((abdômen)))

Reviewer: (((abs)))

Esta referência é para a função interna do Julia para calcular o valor absoluto.

end
----

> Esta função funciona bem para a maior parte dos valores de a, mas em geral é perigoso testar igualdade com pontos flutuantes. Pontos flutuantes não são totalmente corretos: a maioria dos números racionais, como latexmath:[(\frac{1}{3})], e números irracionais, como latexmath:[(\sqrt 2)], não podem ser representados exatamente com um tipo +Float64+.

Reviewer: ```suggestion
Esta função funciona bem para a maior parte dos valores de a, mas em geral é perigoso testar igualdade com pontos flutuantes. Pontos flutuantes não são totalmente corretos: a maioria dos números racionais, como latexmath:[\(\frac{1}{3}\)], e números irracionais, como latexmath:[\(\sqrt 2\)], não podem ser representados exatamente com um tipo +Float64+.
```

end
----

Onde +ε+ (*+\varepsilon TAB+*) possui um valor como +0.0000001+ que determina o quão suficientemente próximo está.

=== Algoritmos

> O Método de Newton é um exemplo de um _algoritmo_: Um processo mecânico para resolver uma categoria de problemas (nesse caso, o cálculo de raízes quadradas).

Reviewer: ```suggestion
O método de Newton é um exemplo de um _algoritmo_: um processo mecânico para resolver uma categoria de problemas (nesse caso, o cálculo de raízes quadradas).
```

(((algoritmo)))

> Para entender o que é um algoritmo, talvez seja interessante mostrar algo que não é um algoritmo. Quando você aprendeu a multiplicar unidades, você provavelmente memorizou a tabuada. De fato, você memorizou 100 soluções específicas. Esse tipo de conhecimento não é um algoritmo.

Reviewer: ```suggestion
Para entender o que é um algoritmo, talvez seja interessante começar com algo que não é um algoritmo. Quando você aprendeu a multiplicar unidades, você provavelmente memorizou a tabuada. De fato, você memorizou 100 soluções específicas. Esse tipo de conhecimento não é um algoritmo.
```

Mas se você fosse "preguiçoso", talvez tivesse aprendido alguns truques. Por exemplo, para encontrar o produto de latexmath:[\(n\)] e 9, você pode escrever latexmath:[\(n-1\)] no primeiro dígito e latexmath:[\(10-n\)] no segundo dígito. Esse truque é uma solução geral para multiplicar qualquer unidade por 9. Isso é um algoritmo!

Similarmente, as técnicas que você aprendeu para a adição com transporte de unidades, a subtração com empréstimos e a divisão longa são todos algoritmos. Uma das características dos algoritmos é que eles não exigem nenhuma inteligência para serem executados. São processos mecânicos em que cada passo segue a partir do último, de acordo com um conjunto simples de regras.

Apesar da execução de algoritmos ser chata, a construção é interessante, intelectualmente desafiadora e uma parte central da ciência da computação.

> Algumas das coisas que as pessoas fazem naturalmente, sem dificuldade ou conscientemente pensado, são as mais difíceis de expressar por algoritmos. Compreender a linguagem nativa é um bom exemplo. Todos fazemos isso, mas até agora ninguém foi capaz de explicar como fazemos, pelo menos não na forma de um algoritmo.

Reviewer: ```suggestion
Algumas das coisas que as pessoas fazem naturalmente, sem dificuldade ou conscientemente pensado, são as mais difíceis de expressar por algoritmos. Compreender a linguagem natural é um bom exemplo. Todos nós fazemos isso, mas até agora ninguém foi capaz de explicar _como_ fazemos, pelo menos não na forma de um algoritmo.
```

> === Depurando

Reviewer: ```suggestion
=== Depuração
```

> Ao começar a escrever programas maiores, você pode passar mais tempo depurando. Mais código significa mais chances de cometer um erro e mais lugares para os erros se esconderem.

Reviewer: ```suggestion
Ao começar a escrever programas maiores, você pode passar mais tempo com a depuração. Mais código significa mais chances de cometer um erro e mais lugares para os erros se esconderem.
```

(((depuração)))

> Uma maneira de reduzir o tempo da depuração é “depurar por bissecção”. Por exemplo, se houver 100 linhas no seu programa e você verificá-las uma de cada vez, serão necessárias 100 etapas.

Reviewer: ```suggestion
Uma maneira de reduzir o tempo da depuração é a “depuração por bissecção”. Por exemplo, se houver 100 linhas no seu programa e você verificá-las uma de cada vez, serão necessárias 100 etapas.
```

> (((depurar por bissecção)))

Reviewer: ```suggestion
(((depuração por bissecção)))
```

Em vez disso, tente quebrar o problema ao meio. Olhe no meio do programa, ou por perto, para um valor intermediário que você pode verificar. Adicione um comando print (ou qualquer outra coisa que tenha um propósito de verificação) e execute o programa.
(((comando print)))

Se a verificação da região do meio estiver incorreta, deve haver um problema na primeira metade do programa. Se estiver correta, o problema está na segunda metade.

> Toda vez que você executa uma verificação como essa, reduz-se pela metade o número de linhas que precisa averiguar. Após seis etapas (que é menor que 100), você reduziria para uma ou duas linhas de código, pelo menos em teoria.

Reviewer: ```suggestion
Toda vez que você executa uma verificação como essa, reduz-se pela metade o número de linhas que se precisa averiguar. Após seis etapas (que é menor que 100), você reduziria para uma ou duas linhas de código, pelo menos em teoria.
```

> Na prática, nem sempre é claro onde é o "meio do programa" e nem sempre é possível verificá-lo. Não faz sentido contar linhas e encontrar o ponto médio exato. Em vez disso, pense nos locais do programa em que pode haver erros e nos locais onde é fácil fazer uma verificação. Em seguida, escolha um local onde você acha que as chances são as mesmas de que o bug seja antes ou depois da verificação.

Reviewer: ```suggestion
Na prática, nem sempre é claro onde é o "meio do programa" e nem sempre é possível verificá-lo. Não faz sentido contar linhas e encontrar o ponto médio exato. Em vez disso, pense nos locais do programa em que pode haver erros e nos locais onde é fácil fazer uma verificação. Em seguida, escolha um local onde você acha que as chances são as mesmas de que o erro seja antes ou depois da verificação.
```

=== Glossário

reatribuição::
Atribuindo um novo valor a uma variável que já existe.
(((reatribuição)))

atualização::
Uma atribuição em que o novo valor da variável depende do antigo.
(((atualização)))

inicialização::
Uma atribuição que fornece um valor inicial a uma variável que será atualizada.
(((inicialização)))

incremento::

> Uma atualização que aumenta o valor de uma variável (geralmente em um).

Reviewer: ```suggestion
Uma atualização que aumenta o valor de uma variável (frequentemente em um).
```

(((incremento)))

decremento::
Uma atualização que diminui o valor de uma variável.
(((decremento)))

iteração::
Execução repetida de um conjunto de comandos usando uma chamada de função recursiva ou um laço.
(((iteração)))

> comando while::

Reviewer: ```suggestion
declaração while::
```

Comando que permite iterações controladas por uma condição.

> (((comando while)))

Reviewer: ```suggestion
(((declaração while)))
```

> comando break::

Reviewer: ```suggestion
declaração break::
```

Comando que permite saltar fora de um laço.

> (((comando break)))

Reviewer: ```suggestion
(((declaração break)))
```

> comando continue::

Reviewer: ```suggestion
declaração continue::
```

Comando dentro de um laço que salta para o início do laço da próxima iteração.

> (((comando continue)))

Reviewer: ```suggestion
(((declaração continue)))
```

laço infinito::
Um laço no qual a sua condição de parada nunca é satisfeita.
(((laço infinito)))

algoritmo::
Um processo geral para resolver uma categoria de problemas.
(((algoritmo)))

=== Exercícios

[[ex07-1]]
===== Exercício 7-2

> Copie o laço de <<square_roots>> e encapsule-o em uma função chamada +minha_raíz+ que usa a variável +a+ como um parâmetro e escolha um valor razoável de +x+ que retorne uma estimativa da raíz quadrada de +a+.

Reviewer: ```suggestion
Copie o laço de <<square_roots>> e encapsule-o em uma função chamada +minha_raiz+ que usa a variável +a+ como um parâmetro e escolha um valor razoável de +x+ que retorne uma estimativa da raíz quadrada de +a+.
```

> Copie o laço de <<square_roots>> e encapsule-o em uma função chamada +minha_raiz+ que usa a variável +a+ como um parâmetro e escolha um valor razoável de +x+ que retorne uma estimativa da raíz quadrada de +a+.

Reviewer: ```suggestion
Copie o laço de <<square_roots>> e encapsule-o em uma função chamada +minha_raiz+ que usa a variável +a+ como um parâmetro e escolha um valor razoável de +x+ que retorne uma estimativa da raiz quadrada de +a+.
```

> (((minha_raíz)))((("função", "definido pelo programador", "minha_raíz", see="minha_raíz")))

Reviewer: ```suggestion
(((minha_raiz)))((("função", "definido pelo programador", "minha_raiz", see="minha_raiz")))
```

> Para testá-la, escreva uma função chamada +testa_raíz+ que imprime uma tabela como esta:

Reviewer: ```suggestion
Para testá-la, escreva uma função chamada +avalia_raiz_quadrada+ que imprime uma tabela como esta:
```

> (((testa_raíz)))((("função", "definido pelo programador", "testa_raíz", see="testa_raíz")))

Reviewer: ```suggestion
(((avalia_raiz_quadrada)))((("função", "definido pelo programador", "avalia_raiz_quadrada", see="avalia_raiz_quadrada")))
```

> [source, @ julia-eval]

Reviewer 2: ```suggestion
[source,@julia-eval]
```

----
using ThinkJulia
io = IOBuffer()

> testa_raíz(io)

Reviewer: ```suggestion
avalia_raiz_quadrada(io)
```

res = String(take!(io))
println(res)
----

> A primeira coluna é um número, +a+; a segunda coluna é a raíz quadrada de +a+ calculado com +minha_raíz+; a terceira coluna é a raíz quadrada calculada por +sqrt+; a quarta coluna é o valor absoluto da diferença entre as duas estimativas.

Reviewer: ```suggestion
A primeira coluna é um número, +a+; a segunda coluna é a raíz quadrada de +a+ calculado com +minha_raiz+; a terceira coluna é a raiz quadrada calculada por +sqrt+; a quarta coluna é o valor absoluto da diferença entre as duas estimativas.
```

> A primeira coluna é um número, +a+; a segunda coluna é a raíz quadrada de +a+ calculado com +minha_raiz+; a terceira coluna é a raiz quadrada calculada por +sqrt+; a quarta coluna é o valor absoluto da diferença entre as duas estimativas.

Reviewer: ```suggestion
A primeira coluna é um número, +a+; a segunda coluna é a raiz quadrada de +a+ calculado com +minha_raiz+; a terceira coluna é a raiz quadrada calculada por +sqrt+; a quarta coluna é o valor absoluto da diferença entre as duas estimativas.
```

[[ex07-2]]
===== Exercício 7-3

> A função interna +Meta.parse+ pega uma string e transforma-a em uma expressão. Essa expressão pode ser avaliada em Julia com a função +Core.eval+. Por exemplo:

Reviewer: ```suggestion
A função interna +Meta.parse+ recebe uma string e transforma-a em uma expressão. Essa expressão pode ser avaliada em Julia com a função +Core.eval+. Por exemplo:
```

> (((parse)))((("função","Meta","parse", see="parse")))(((eval)))((("função", "Core", "eval", see="eval")))

Reviewer: ```suggestion
(((parse)))((("função", "Meta", "parse", see="parse")))(((eval)))((("função", "Core", "eval", see="eval")))
```

end
----

> Onde +ε+ (*+\varepsilon TAB+*) possui um valor como +0.0000001+ que determina o quão próximo está.

Reviewer: "que determina o quão _suficientemente_ próximo está."

=== Algoritmos

> O Método de Newton é um exemplo de um _algoritmo_: É um processo mecânico de se resolver uma categoria de problemas (neste caso, raízes quadradas).

Reviewer: "É um processo mecânico _para_ resolver uma categoria de problemas (neste caso, _o cálculo de_ raízes quadradas)."

> (((algoritmos)))

Reviewer: (((algoritmo)))

> Para entender o que é um algoritmo, ajudaria começar com algo que não é um algoritmo. Quando você aprendeu a multiplicar unidades, você provavelmente memorizou a tabuada. Com efeito, você memorizou 100 soluções específicas. Esse tipo de conhecimento não é um algoritmo.

Reviewer: "_De fato_, você memorizou 100 soluções específicas."

> Mas se você fosse "preguiçoso", talvez tivesse aprendido alguns truques. Por exemplo, para encontrar o produto de latexmath:[\(n\)] e 9, você escreve latexmath:[\(n-1\)] no primeiro dígito e latexmath:[\(10-n\)] no segundo dígito. Esse truque é uma solução geral para multiplicar qualquer unidade por 9. Isso é um algoritmo!

Reviewer: ```suggestion
Mas se você fosse "preguiçoso", talvez tivesse aprendido alguns truques. Por exemplo, para encontrar o produto de latexmath:[\(n\)] e 9, você pode escrever latexmath:[\(n-1\)] no primeiro dígito e latexmath:[\(10-n\)] no segundo dígito. Esse truque é uma solução geral para multiplicar qualquer unidade por 9. Isso é um algoritmo!
```

> Similarmente, as técnicas que você aprendeu para adição com transporte de unidades, subtração com empréstimos e divisão longa são todos algoritmos. Uma das características dos algoritmos é que eles não exigem nenhuma inteligência para serem executados. São processos mecânicos em que cada etapa segue da última, de acordo com um conjunto simples de regras.

Reviewer: Minha sugestão: "Similarmente, as técnicas que você aprendeu para _a_ adição com transporte de unidades, _a_ subtração com empréstimos e _a_ divisão longa são todos algoritmos. ... São processos mecânicos em que cada etapa segue _a partir_ da última, de acordo com um conjunto simples de regras."

Na minha opinião, essa sugestão dá mais fluidez para a frase.

> Similarmente, as técnicas que você aprendeu para a adição com transporte de unidades, a subtração com empréstimos e a divisão longa são todos algoritmos. Uma das características dos algoritmos é que eles não exigem nenhuma inteligência para serem executados. São processos mecânicos em que cada etapa segue a partir da última, de acordo com um conjunto simples de regras.

Reviewer: ```suggestion
Similarmente, as técnicas que você aprendeu para a adição com transporte de unidades, a subtração com empréstimos e a divisão longa são todos algoritmos. Uma das características dos algoritmos é que eles não exigem nenhuma inteligência para serem executados. São processos mecânicos em que cada passo segue a partir do último, de acordo com um conjunto simples de regras.
```

> A execução de algoritmos é chata, mas projetá-los é interessante, intelectualmente desafiadora e parte central da ciência da computação.

Reviewer: "A execução de algoritmos é chata, mas _elaborá_-los é interessante, intelectualmente _desafiador_ e _uma_ parte central da ciência da computação."

> Apesar da execução de algoritmos seja chata, a construção é interessante, intelectualmente desafiadora e uma parte central da ciência da computação.

Reviewer: ```suggestion
Apesar da execução de algoritmos ser chata, a construção é interessante, intelectualmente desafiadora e uma parte central da ciência da computação.
```

> Algumas das coisas que as pessoas fazem naturalmente, sem dificuldade ou pensamento consciente, são as mais difíceis de expressar por algoritmos. Compreender a linguagem natural é um bom exemplo. Todos fazemos isso, mas até agora ninguém foi capaz de explicar como o fazemos, pelo menos não na forma de um algoritmo.

Reviewer: "Todos fazemos isso, mas até agora ninguém foi capaz de explicar como fazemos, pelo menos não na forma de um algoritmo."

> Algumas das coisas que as pessoas fazem naturalmente, sem dificuldade ou pensamento consciente, são as mais difíceis de expressar por algoritmos. Compreender a linguagem natural é um bom exemplo. Todos fazemos isso, mas até agora ninguém foi capaz de explicar como fazemos, pelo menos não na forma de um algoritmo.

Reviewer: ```suggestion
Algumas das coisas que as pessoas fazem naturalmente, sem dificuldade ou conscientemente pensado, são as mais difíceis de expressar por algoritmos. Compreender a linguagem natural é um bom exemplo. Todos fazemos isso, mas até agora ninguém foi capaz de explicar como fazemos, pelo menos não na forma de um algoritmo.
```

> === Debugando

Reviewer: ```suggestion
=== Depuração
```

> Ao começar a escrever programas maiores, você pode passar mais tempo depurando. Mais código significa mais chances de cometer um erro e mais lugares para ocultar os erros.

Reviewer: "Ao começar a escrever programas maiores, você pode passar mais tempo _debugando_. Mais código significa mais chances de cometer um erro e mais lugares para _os erros se esconderem_."

> Ao começar a escrever programas maiores, você pode passar mais tempo debugando. Mais código significa mais chances de cometer um erro e mais lugares para os erros se esconderem.

Reviewer: ```suggestion
Ao começar a escrever programas maiores, você pode passar mais tempo depurando. Mais código significa mais chances de cometer um erro e mais lugares para os erros se esconderem.
```

> (((debugando)))

Reviewer: ```suggestion
(((depuração)))
```

> Uma maneira de reduzir o tempo de debug é "debug por bissecção". Por exemplo, se houver 100 linhas no seu programa e você as verificar uma de cada vez, serão necessárias 100 etapas.

Reviewer: "Uma maneira de reduzir o tempo _do debugging_ é o “debugging por bissecção”. Por exemplo, se houver 100 linhas no seu programa e você _verificá-las_ uma de cada vez, serão necessárias 100 etapas."

> (((debug por bissecção)))

Reviewer: (((debugging por bissecção)))

> Uma maneira de reduzir o tempo do debugging é o “debugging por bissecção”. Por exemplo, se houver 100 linhas no seu programa e você verificá-las uma de cada vez, serão necessárias 100 etapas.

Reviewer: ```suggestion
Uma maneira de reduzir o tempo da depuração é a “depuração por bissecção”. Por exemplo, se houver 100 linhas no seu programa e você verificá-las uma de cada vez, serão necessárias 100 etapas.
```

> (((debugging por bissecção)))

Reviewer: ```suggestion
(((depuração por bissecção)))
```

> Em vez disso, tente quebrar o problema ao meio. Olhe no meio do programa, ou perto dele, para um valor intermediário que você pode verificar. Adicione uma atribuição de impressão (ou outra coisa que tenha um efeito verificável) e execute o programa.

Reviewer: "Adicione _um comando print_ (ou _qualquer_ outra coisa que tenha um _propósito de verificação_) e execute o programa."

> (((atribuição de impressão)))

Reviewer: (((comando print)))

> Em vez disso, tente quebrar o problema ao meio. Olhe no meio do programa, ou perto dele, para um valor intermediário que você pode verificar. Adicione um comando print (ou qualquer outra coisa que tenha um propósito de verificação) e execute o programa

Reviewer: ```suggestion
Em vez disso, tente quebrar o problema ao meio. Olhe no meio do programa, ou perto dele, para um valor intermediário que você pode verificar. Adicione um comando print (ou qualquer outra coisa que tenha um propósito de verificação) e execute o programa.
```

> Se a verificação do ponto médio estiver incorreta, deve haver um problema na primeira metade do programa. Se estiver correto, o problema está no segundo semestre.

Reviewer: "Se a verificação _da região do meio_ estiver incorreta, deve haver um problema na primeira metade do programa. Se estiver correto, o problema está _na segunda metade_."

> Toda vez que você executa uma verificação como essa, reduz pela metade o número de linhas que precisa pesquisar. Após seis etapas (que é menor que 100), você reduziria para uma ou duas linhas de código, pelo menos em teoria.

Reviewer: "Toda vez que você executa uma verificação como essa, reduz_-se_ pela metade o número de linhas que precisa averiguar. Após seis etapas (que é menor que 100), você reduziria para uma ou duas linhas de código, pelo menos em teoria."

> Toda vez que você executa uma verificação como essa, reduz_-se_ pela metade o número de linhas que precisa averiguar. Após seis etapas (que é menor que 100), você reduziria para uma ou duas linhas de código, pelo menos em teoria.

Reviewer: ```suggestion
Toda vez que você executa uma verificação como essa, reduz-se pela metade o número de linhas que precisa averiguar. Após seis etapas (que é menor que 100), você reduziria para uma ou duas linhas de código, pelo menos em teoria.
```

> Na prática, nem sempre é claro qual é o "meio do programa" e nem sempre é possível verificá-lo. Não faz sentido contar linhas e encontrar o ponto médio exato. Em vez disso, pense nos locais do programa em que pode haver erros e nos locais onde é fácil fazer uma verificação. Em seguida, escolha um local onde você acha que as chances são as mesmas de que o bug seja antes ou depois da verificação.

Reviewer: "Na prática, nem sempre é claro _onde_ é o "meio do programa" ..."

=== Glossário

reatribuição::
Atribuindo um novo valor a uma variável que já existe.
(((reatribuição)))

> atualizar::

Reviewer: Na minha opinião, atualização é melhor que atualizar.

Uma atribuição em que o novo valor da variável depende do antigo.
(((atualizar)))

inicialização::
Uma atribuição que fornece um valor inicial a uma variável que será atualizada.
(((inicialização)))

incremento::

> Uma atualização que aumenta o valor de uma variável (geralmente em uma).

Reviewer: "Uma atualização que aumenta o valor de uma variável (geralmente em _um_)."

(((incremento)))

decremento::
Uma atualização que diminui o valor de uma variável.

> ((decremento)))

Reviewer: São três parênteses no começo e três no final.

iteração::

> Execução repetida de um conjunto de instruções usando uma chamada de função recursiva ou um laço.

Reviewer: "Execução repetida de um conjunto de comandos usando uma chamada de função recursiva ou um laço."

(((iteração)))

> atribuição while::

Reviewer: "Declaração while".

> Instrução que permite iterações controladas por uma condição.

Reviewer: "_Declaração_ que permite iterações controladas por uma condição."

(((declaração while)))

> atribuição de parada::

Reviewer: "Declaração break"

Aproveita e já corrige linha 380.

> Instrução que permite saltar fora de um laço.

Reviewer: "Declaração que permite saltar fora de um laço."

(((atribuição de parada)))

> atribuição de continuidade::

Reviewer: "Declaração continue"

Aproveita e já corrige linha 384.

> Instrução dentro de um laço que salta para o início do laço para a próxima iteração.

Reviewer: "Declaração dentro de um laço que salta para o início do laço da próxima iteração."

> declaração dentro de um laço que salta para o início do laço para a próxima iteração.

Reviewer: ```suggestion
Declaração dentro de um laço que salta para o início do laço da próxima iteração.
```

(((atribuição de continuidade)))

laço infinito::

> Um laço no qual a condição final nunca é satisfeita.

Reviewer: Minha sugestão: "Um laço no qual a sua condição nunca é satisfeita."

Na minha opinião, esta sugestão deixa a frase mais específica.

> Um laço no qual a sua condição nunca é satisfeita.

Reviewer: ```suggestion
Um laço no qual a sua condição de parada nunca é satisfeita.
```

(((laço infinito)))

algoritmo::
Um processo geral para resolver uma categoria de problemas.
(((algoritmo)))

=== Exercícios

[[ex07-1]]
===== Exercício 7-2

> Copie o laço de <<raízes_quadradas>> e encapsule-o em uma função chamada +minha_raiz+ que use +a+ como parâmetro, escolha um valor razoável de +x+ e retorne uma estimativa da raiz quadrada de +a+.

Reviewer: Copie o laço de <<raízes_quadradas>> e encapsule-o em uma função chamada +minha_raiz+ que use a variável +a+ como parâmetro, _e_ escolha um valor razoável de +x+ _que_ retorne uma estimativa da raiz quadrada de +a+.

> Copie o laço de <<square_roots>> e encapsule-o em uma função chamada +minha_raiz+ que use a variável +a+ como parâmetro, e escolha um valor razoável de +x+ que retorne uma estimativa da raiz quadrada de +a+.

Reviewer: ```suggestion
Copie o laço de <<square_roots>> e encapsule-o em uma função chamada +minha_raiz+ que use a variável +a+ como um parâmetro, e escolha um valor razoável de +x+ que retorne uma estimativa da raiz quadrada de +a+.
```

> (((minha_raiz))) ((("função", "definido pelo programador", "minha_raiz", see = "minha_raiz"))))

Reviewer: (((minha_raiz)))((("função", "definido pelo programador", "minha_raiz", see="minha_raiz")))

> (((minha_raiz)))((("função", "definido pelo programador", "minha_raiz", see = "minha_raiz"))))

Reviewer: ```suggestion
(((minha_raiz)))((("função", "definido pelo programador", "minha_raiz", see="minha_raiz")))
```

> Para testá-lo, escreva uma função chamada +testa_raiz+ que imprime uma tabela como esta:

Reviewer: "Para _testá-la_..."

> (((testa_raiz))) ((("função", "definido pelo programador", "testa_raiz", see = "testa_raiz"))))

Reviewer: (((testa_raiz)))((("função", "definido pelo programador", "testa_raiz", see="testa_raiz")))

> (((testa_raiz)))((("função", "definido pelo programador", "testa_raiz", see="testa_raiz"))))

Reviewer: ```suggestion
(((testa_raiz)))((("função", "definido pelo programador", "testa_raiz", see="testa_raiz")))
```

> [fonte, @ julia-eval]

Reviewer: ```suggestion
[source, @ julia-eval]
```

----
using ThinkJulia
io = IOBuffer()
testa_raiz(io)

> fora = String(take!(io))

Reviewer: Eu acho q o "out" é de output (resultado). Se sim, é melhor traduzir out para "res".

println(fora)
----

> A primeira coluna é um número, +a+; a segunda coluna é a raiz quadrada de um calculado com +minha_raiz+; a terceira coluna é a raiz quadrada calculada por +sqrt+; a quarta coluna é o valor absoluto da diferença entre as duas estimativas.

Reviewer: "A primeira coluna é _o número_ +a+; a segunda coluna é a raiz quadrada de _a_ calculado com +minha_raiz+; ..."

[[ex07-2]]
===== Exercício 7-3

> A função interna +Meta.parse+ pega uma string e a transforma em uma expressão. Essa expressão pode ser avaliada em Julia com a função +Core.eval+. Por exemplo:

Reviewer: ```suggestion
A função interna +Meta.parse+ pega uma string e transforma-a em uma expressão. Essa expressão pode ser avaliada em Julia com a função +Core.eval+. Por exemplo:
```

> (((análise)))((("função","Meta","análise", see="análise"))) (((eval))) ((("function", "Core", "eval", see="eval")))

Reviewer: (((parse)))((("função", "Meta", "parse", see="parse")))

> (((parse)))((("função","Meta","parse", see="parse"))) (((eval)))((("function", "Core", "eval", see="eval")))

Reviewer: ```suggestion
(((parse)))((("função","Meta","parse", see="parse")))(((eval)))((("função", "Core", "eval", see="eval")))
```

> [fonte, @ julia-eval chap07]

Reviewer: ```suggestion
[source,@julia-eval chap07]
```

----
import Base.eval
----

> [fonte, @ julia-repl-test chap07]

Reviewer: ```suggestion
[source,@julia-repl-test chap07]
```

----
julia> expr = Meta.parse ("1 + 2 * 3")
: (1 + 2 * 3)
julia> eval (expr)
7
julia> expr = Meta.parse ("sqrt (π)")
: (sqrt (π))

> julia> eval (expr)

Reviewer: Arrumar esses espaços adicionais colocado no código de Julia da linha 430 até linha 436.

1.7724538509055159
----

> Escreva uma função chamada +avalie_laço+ que solicite iterativamente ao usuário, pegue a entrada resultante e a avalie usando +eval+ e imprima o resultado. Ele deve continuar até o usuário digitar +concluído+ e retornar o valor da última expressão avaliada.

Reviewer: "Escreva uma função chamada +avalie_laço+ que solicite iterativamente ao usuário, _e então_ pegue a entrada _recebida_ e a avalie usando +eval+ e _depois_ imprima o resultado. Ele deve continuar até o usuário digitar +concluído+ e retornar o valor da última expressão avaliada."

> Escreva uma função chamada +avalie_laço+ que solicite iterativamente ao usuário, e então pegue a entrada recebida, avalie-na usando +eval+ e depois imprima o resultado. Ele deve continuar até o usuário digitar +concluído+ e retornar o valor da última expressão avaliada.

Reviewer: ```suggestion
Escreva uma função chamada +avalie_laço+ que solicite iterativamente ao usuário, e então pegue a entrada recebida, avalie-na usando +eval+ e depois imprima o resultado. Ele deve continuar até o usuário digitar +concluído+ e depois retornar o valor da última expressão avaliada.
```

(((avalie_laço)))((("função", "definido pelo programador", "avalie_laço", see= "avalie_laço")))

[[ex07-3]]
===== Exercício 7-4

> O matemático Srinivasa Ramanujan encontrou uma série infinita que pode ser usada para gerar uma aproximação numérica do latexmath: [\(\ frac {1}{\ pi}\)]:

Reviewer: "para gerar uma aproximação numérica _de_ latexmath: [\(\ frac {1}{\ pi}\)]:"

=== Algoritmos

O Método de Newton é um exemplo de um _algoritmo_: É um processo mecânico para resolver uma categoria de problemas (neste caso, o cálculo de raízes quadradas).
(((algoritmo)))

Para entender o que é um algoritmo, ajudaria começar com algo que não é um algoritmo. Quando você aprendeu a multiplicar unidades, você provavelmente memorizou a tabuada. De fato, você memorizou 100 soluções específicas. Esse tipo de conhecimento não é um algoritmo.

Mas se você fosse "preguiçoso", talvez tivesse aprendido alguns truques. Por exemplo, para encontrar o produto de latexmath:[\(n\)] e 9, você escreve latexmath:[\(n-1\)] no primeiro dígito e latexmath:[\(10-n\)] no segundo dígito. Esse truque é uma solução geral para multiplicar qualquer unidade por 9. Isso é um algoritmo!

Similarmente, as técnicas que você aprendeu para a adição com transporte de unidades, a subtração com empréstimos e a divisão longa são todos algoritmos. Uma das características dos algoritmos é que eles não exigem nenhuma inteligência para serem executados. São processos mecânicos em que cada etapa segue a partir da última, de acordo com um conjunto simples de regras.

> A execução de algoritmos é chata, mas elaborá-los é interessante, intelectualmente desafiador e uma parte central da ciência da computação.

Reviewer: Sugestão para deixar a frase "mais poderosa":
```suggestion
Apesar da execução de algoritmos seja chata, a construção é interessante, intelectualmente desafiadora e uma parte central da ciência da computação.
```

Mas se você fosse "preguiçoso", talvez tivesse aprendido alguns truques. Por exemplo, para encontrar o produto de latexmath:[\(n\)] e 9, você pode escrever latexmath:[\(n-1\)] no primeiro dígito e latexmath:[\(10-n\)] no segundo dígito. Esse truque é uma solução geral para multiplicar qualquer unidade por 9. Isso é um algoritmo!

Similarmente, as técnicas que você aprendeu para a adição com transporte de unidades, a subtração com empréstimos e a divisão longa são todos algoritmos. Uma das características dos algoritmos é que eles não exigem nenhuma inteligência para serem executados. São processos mecânicos em que cada passo segue a partir do último, de acordo com um conjunto simples de regras.

Apesar da execução de algoritmos ser chata, a construção é interessante, intelectualmente desafiadora e uma parte central da ciência da computação.

Algumas das coisas que as pessoas fazem naturalmente, sem dificuldade ou conscientemente pensado, são as mais difíceis de expressar por algoritmos. Compreender a linguagem nativa é um bom exemplo. Todos fazemos isso, mas até agora ninguém foi capaz de explicar como fazemos, pelo menos não na forma de um algoritmo.

> === Depuração

Author: ```suggestion
=== Depurando
```

Author: Apenas alterando 'Debugging" para "Depurando".

Reviewer: @abelsiqueira , resolve aí.

Tanto a tradução espanhol como o Pense Python em português está escrito depuração.

Author: Então vamos manter "Depuração".

Reviewer 2: @chunggnuhc, eu te coloquei como Reviewer de #29, que é esse assunto.

Reviewer: @abelsiqueira mandou deixar "Depurando".

Atribuindo um novo valor a uma variável que já existe.
(((reatribuição)))

atualização::
Uma atribuição em que o novo valor da variável depende do antigo.
(((atualização)))

inicialização::
Uma atribuição que fornece um valor inicial a uma variável que será atualizada.
(((inicialização)))

incremento::
Uma atualização que aumenta o valor de uma variável (geralmente em um).
(((incremento)))

decremento::
Uma atualização que diminui o valor de uma variável.
(((decremento)))

iteração::
Execução repetida de um conjunto de comandos usando uma chamada de função recursiva ou um laço.
(((iteração)))

declaração while::

> declração que permite iterações controladas por uma condição.

Reviewer: ```suggestion
Declaração que permite iterações controladas por uma condição.
```

(((declaração while)))

declaração break::

> declração que permite saltar fora de um laço.

Reviewer: ```suggestion
Declaração que permite saltar fora de um laço.
```

[[ex07-1]]
===== Exercício 7-2

> Copie o laço de <<square_roots>> e encapsule-o em uma função chamada +minha_raiz+ que use +a+ como parâmetro, e escolha um valor razoável de +x+ e retorne uma estimativa da raiz quadrada de +a+.

Reviewer: ```suggestion
Copie o laço de <<square_roots>> e encapsule-o em uma função chamada +minha_raiz+ que use a variável +a+ como parâmetro, e escolha um valor razoável de +x+ que retorne uma estimativa da raiz quadrada de +a+.
```

(((minha_raiz)))((("função", "definido pelo programador", "minha_raiz", see = "minha_raiz"))))

Para testá-la, escreva uma função chamada +testa_raiz+ que imprime uma tabela como esta:

> (((testa_raiz)))((("função", "definido pelo programador", "testa_raiz", see = "testa_raiz"))))

Reviewer: ```suggestion
(((testa_raiz)))((("função", "definido pelo programador", "testa_raiz", see="testa_raiz"))))
```

1.7724538509055159
----

> Escreva uma função chamada +avalie_laço+ que solicite iterativamente ao usuário, e então pegue a entrada recebida e a avalie usando +eval+ e depois imprima o resultado. Ele deve continuar até o usuário digitar +concluído+ e retornar o valor da última expressão avaliada.

Reviewer: ```suggestion
Escreva uma função chamada +avalie_laço+ que solicite iterativamente ao usuário, e então pegue a entrada recebida, avalie-na usando +eval+ e depois imprima o resultado. Ele deve continuar até o usuário digitar +concluído+ e retornar o valor da última expressão avaliada.
```

1.7724538509055159
----

> Escreva uma função chamada +avalie_laço+ que solicite iterativamente ao usuário e então pegue a entrada recebida, avalie-na usando +eval+ e depois imprima o resultado. A função deve continuar até o usuário digitar +concluído+ e depois retornar o valor da última expressão avaliada.

Reviewer: ```suggestion
Escreva uma função chamada +avalie_laço+ que solicite iterativamente ao usuário, pegue a entrada recebida e avalie-a usando +eval+ e depois imprime o resultado. A função deve continuar até o usuário digitar +concluído+ e depois retornar o valor da última expressão avaliada.
```

\end{equation}
++++

> Escreva uma função chamada +estimativa_pi+ que use essa fórmula para calcular e retornar uma estimativa de π. Ele deve usar um laço while para calcular os termos da soma até que o último termo seja menor que +1e-15+ (que é a notação Julia para latexmath:[(10^{-15}\)]). Você pode verificar o resultado comparando-o com +π+.

Reviewer: ```suggestion
Escreva uma função chamada +estimativa_pi+ que use essa fórmula para calcular e retornar uma estimativa de π. Ele deve usar um laço while para calcular os termos da soma até que o último termo seja menor que +1e-15+ (que é a notação Julia para latexmath:[\(10^{-15}\)]). Você pode verificar o resultado comparando-o com +π+.
```

> (((estimativa_pi)))((("função", "definido pelo programador", "estimativa_pi", see = "estimativa_pi"))))

Reviewer: ```suggestion
(((estimativa_pi)))((("função", "definido pelo programador", "estimativa_pi", see="estimativa_pi"))))
```

\end{equation}
++++

> Escreva uma função chamada +estimativa_pi+ que use essa fórmula para calcular e retornar uma estimativa de π. Ele deve usar um laço while para calcular os termos da soma até que o último termo seja menor que +1e-15+ (que é a notação Julia para latexmath:[\(10^{-15}\)]). Você pode verificar o resultado comparando-o com +π+.

Reviewer: ```suggestion
Escreva uma função chamada +estimativa_pi+ que use essa fórmula para calcular e retornar uma estimativa de π. Ele deve usar um laço while para calcular os termos da soma até que o último termo seja menor que +1e-15+ (que é a notação do Julia para latexmath:[\(10^{-15}\)]). Você pode verificar o resultado comparando-o com +π+.
```

> Escreva uma função chamada +estimativa_pi+ que use essa fórmula para calcular e retornar uma estimativa de π. Ele deve usar um laço while para calcular os termos da soma até que o último termo seja menor que + 1e-15 + (que é a notação Julia para o latexmath:[\(10^{-15}\)]). Você pode verificar o resultado comparando-o com +π+.

Reviewer: "Ele deve usar um laço while para calcular os termos da soma até que o último termo seja menor que +1e-15+ (que é a notação Julia para latexmath:[\(10^{-15}\)])."

> Escreva uma função chamada +estimativa_pi+ que use essa fórmula para calcular e retornar uma estimativa de π. Ele deve usar um laço while para calcular os termos da soma até que o último termo seja menor que +1e-15+ (que é a notação do Julia para latexmath:[\(10^{-15}\)]). Você pode verificar o resultado comparando-o com +π+.
> (((estimativa_pi)))((("função", "definido pelo programador", "estimativa_pi", see="estimativa_pi"))))

Reviewer 2: ```suggestion
Escreva uma função chamada +estima_pi+ que use essa fórmula para calcular e retornar uma estimativa de π. Ela deve usar um laço while para calcular os termos da soma até que o último termo seja menor que +1e-15+ (que é a notação do Julia para latexmath:[\(10^{-15}\)]). Você pode verificar o resultado comparando-o com +π+.
(((estima_pi)))((("função", "definido pelo programador", "estima_pi", see="estima_pi"))))
```

> (((estimativa pi))) ((("função", "definido pelo programador", "estimativa_pi", see = "estimativa_pi"))))

Reviewer: (((estimativa_pi)))((("função", "definido pelo programador", "estimativa_pi", see="estimativa_pi")))
